python/strings_udf/cpp/include/cudf/strings/udf/udf_string.cuh

```diff
@@ -0,0 +1,475 @@
+/*
+ * Copyright (c) 2020-2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "udf_string.hpp"
+
+#include <cudf/strings/detail/utf8.hpp>
+#include <cudf/strings/string_view.cuh>
+
+#include <algorithm>
+#include <limits>
+#include <string>
+
+namespace cudf {
+namespace strings {
+namespace udf {
+namespace detail {
+
+__device__ inline static cudf::size_type bytes_in_null_terminated_string(char const* str)
+{
+  if (!str) return 0;
+  cudf::size_type bytes = 0;
+  while (*str++)
+    ++bytes;
+  return bytes;
+}
+
+}  // namespace detail
+
+__device__ inline char* udf_string::allocate(cudf::size_type bytes)
```

Comment:
I disagree with this. I know I've caved on this in the past. The function signature declares pass by value and the caller knows the value passed will not change (on their side) but it gives the option to the implementation on how to use the parameter internally. The optimizer is very good at recognizing this correctly and I've seen no evidence that declaring a _parameter_ as `const` will make it faster.  We declare variables as `const` only to communicate to the reader that it will not be changed inside the function. I think that is an internal detail that should not be leaked into the function signature.

Reply:
Fair enough. I have in fact seen performance differences between such cases, but typically only when the call stack is very deep and where the compiler is already overworked. In lieu of any compelling profiling data here your argument makes sense.